src/main/java/org/dcsa/tnt/service/impl/NotificationSignatureHandler.java

```diff
@@ -0,0 +1,374 @@
+                                            } else if (responseCode < 500) {
+                                                statusMessage = "Subscriber rejected message suggesting a client error (HTTP " + responseCode + ")";
+                                            } else {
+                                                statusMessage = "Subscriber was unavailable (HTTP " + responseCode + ")";
+                                            }
+                                            break;
+                                    }
+                                    if (!respectRetryAfter || nextAttempt == null) {
+                                        nextAttempt = computeNextDelay(eventSubscriptionState);
+                                    }
+                                    eventSubscriptionState.setRetryAfter(nextAttempt);
+                                    log.debug("Notification for " + callbackUrl + " failed, it replied with: "
+                                            + responseCode + ": Will retry after " + nextAttempt + " (status: "
+                                            + statusMessage + ")");
+                                }
+                                eventSubscriptionState.setLastStatusMessage(statusMessage);
+                                return Mono.just(eventSubscriptionState);
+                            });
+                });
+    }
+
+    private OffsetDateTime parseRetryAfterHeader(String callbackUrl, String retryAfterRaw) {
+        OffsetDateTime parsedDateTime = null;
+
+        if (!retryAfterRaw.isEmpty() && Character.isDigit(retryAfterRaw.charAt(0))) {
+            try {
+                long delay = Long.parseLong(retryAfterRaw);
+                parsedDateTime = OffsetDateTime.now().plusSeconds(delay);
+            } catch (NumberFormatException e) {
+                log.debug(callbackUrl + " returned invalid Retry-After header (delay-variant); ignoring");
+            }
+        } else {
+            try {
+                parsedDateTime = OffsetDateTime.parse(retryAfterRaw, DateTimeFormatter.RFC_1123_DATE_TIME);
+            } catch (DateTimeParseException e) {
+                log.debug(callbackUrl + " returned invalid Retry-After header (date-variant); ignoring");
+            }
+        }
+        return parsedDateTime;
+    }
+
+    private OffsetDateTime computeNextDelay(EventSubscriptionState eventSubscriptionState) {
+        Long delaySeconds = eventSubscriptionState.getAccumulatedRetryDelay();
+        long limit = notificationServiceConfig.getMaxRetryAfterDelay().toSeconds();
+        if (delaySeconds == null) {
+            delaySeconds = notificationServiceConfig.getMinRetryAfterDelay().toSeconds();
+        } else {
+            delaySeconds *= 2;
+        }
+        if (delaySeconds >= limit) {
+            delaySeconds = limit;
+        }
+        eventSubscriptionState.setAccumulatedRetryDelay(delaySeconds);

```

Comment:
Isn't this dangerous to update? Why not update eventSubscriptionState when you return from this call?